PlatformError: GLX: Failed to make context current

I'm seeing an issue that I've seen before. I know how it can be fixed, but I'm not succeeding in applying the fix to this project.

In flutter-desktop-embedding I encountered an error that was caused by improper resolving of GL functions. The Flutter>Skia>GL rendering would assume the wrong GL context and resolved the wrong procs, ending up in a bunch of errors. Providing flutter with a glfw proc resolver fixed the issue. This occurs on machines with multiple video cards (e.g.: intel graphics as part of the CPU and an nvidia card for when there's some actual GPU power required).

This issue was fixed in flutter-desktop-embedding by the following PR: https://github.com/google/flutter-desktop-embedding/pull/184

I really like Go, have been working with it since before Go v 1.0. So obviously I wanted to try this project as well. I tried to rework the fix over for this project, but ran into a few issues.

First of all, the go-gl/glfw project doesn't expose `glfwGetProcAddress`.
There's an open issue about it: https://github.com/go-gl/glfw/issues/234

I tried to start patching this by using glfw directly. Not as a permanent solution, but just to try out the fix and create a use-case example for go-gl/glfw to expose `glfwGetProcAddress`. ~However, it's not working, compile warning on the proxy function I made, I'm probably doing something wrong in C land.
The changes I made are visible here: https://github.com/Drakirus/go-flutter-desktop-embedder/compare/master...GeertJohan:feature/add-gl-proc-resolver~

~The compile error on my feature branch:~

> ~\# github.com/Drakirus/go-flutter-desktop-embedder/flutter
> /tmp/go-build300793477/b036/_x004.o: In function `_cgo_b8b7151b1d3c_Cfunc_glfwGetProcAddress':
> /tmp/go-build/cgo-gcc-prolog:61: undefined reference to `glfwGetProcAddress'
> /tmp/go-build300793477/b036/_x005.o: In function `runFlutter':
> ../../flutter/flutter_helper.c:31: undefined reference to `proxy_gl_proc_resolver'
> collect2: error: ld returned 1 exit status~

**Edit: The quick'n'dirty fix works now. See comment below.**

Note that I had to copy over some of the new fields and structs of the embedder API to get access to the `gl_proc_resolver` parameter.

I hope we may be able to fix this, preferably by having `glfwGetProcAddress` exposed from the go-gl/glfw package. Any help is welcome.

## Flutter version

```
Flutter 1.2.1 • channel dev • https://github.com/flutter/flutter.git
Framework • revision 8661d8aecd (2 days ago) • 2019-02-14 19:19:53 -0800
Engine • revision 3757390fa4
Tools • Dart 2.1.2 (build 2.1.2-dev.0.0 0a7dcf17eb)

```

## Does the prebuilt and portable version works ?
- No

## Steps to Reproduce

1. Build following the instructions
2. Run the binary (`./main`)

> [WARNING:flutter/shell/platform/embedder/embedder.cc(312)] FlutterProjectArgs.main_path is deprecated and should be set null.
> [WARNING:flutter/shell/platform/embedder/embedder.cc(317)] FlutterProjectArgs.packages_path is deprecated and should be set null.
> [ERROR:flutter/shell/platform/embedder/embedder_surface_gl.cc(89)] Could not create a resource context for async texture uploads. Expect degraded performance. Set a valid make_resource_current callback on FlutterOpenGLRendererConfig.
> [ERROR:flutter/shell/gpu/gpu_surface_gl.cc(85)] Failed to setup Skia Gr context.
> [ERROR:flutter/shell/platform/embedder/embedder_surface_gl.cc(89)] Could not create a resource context for async texture uploads. Expect degraded performance. Set a valid make_resource_current callback on FlutterOpenGLRendererConfig.
> flutter: Observatory listening on http://127.0.0.1:50300/
> 2019/02/16 16:28:59 PlatformError: GLX: Failed to make context current
> [1]    18962 segmentation fault (core dumped)  ./main

--- a/flutter/flutter_helper.c
+++ b/flutter/flutter_helper.c
@@ -9,6 +9,7 @@ bool proxy_clear_current(void *v);
 bool proxy_present(void *v);
 uint32_t proxy_fbo_callback(void *v);
 bool proxy_make_resource_current(void *v);
+void *proxy_gl_proc_resolver(void *v, const char *procname);
 
 bool proxy_on_platform_message(FlutterPlatformMessage *message,
                              void *window);
@@ -27,6 +28,7 @@ FlutterResult runFlutter(uintptr_t window, FlutterEngine *engine, FlutterProject
   config.open_gl.present = proxy_present;
   config.open_gl.fbo_callback = proxy_fbo_callback;
   config.open_gl.make_resource_current = proxy_make_resource_current;
+  config.open_gl.gl_proc_resolver = proxy_gl_proc_resolver;
 
 	Args->command_line_argc = nVmAgrs;
 	Args->command_line_argv = vmArgs;

--- a/flutter/flutter_proxy_glfw.go
+++ b/flutter/flutter_proxy_glfw.go
@@ -70,3 +70,8 @@ func proxy_make_resource_current(v unsafe.Pointer) C.bool {
 	index := *(*C.int)(w.GetUserPointer())
 	return C.bool(flutterEngines[index].FMakeResourceCurrent(v))
 }
+
+//export proxy_gl_proc_resolver
+func proxy_gl_proc_resolver(v unsafe.Pointer, procname *C.char) unsafe.Pointer {
+	return glfw.GetProcAddress(C.GoString(procname))
+}

--- a/flutter/library/flutter_embedder.h
+++ b/flutter/library/flutter_embedder.h
@@ -31,8 +31,31 @@ typedef enum {
 
 typedef struct _FlutterEngine* FlutterEngine;
 
+typedef struct {
+  //   horizontal scale factor
+  double scaleX;
+  //    horizontal skew factor
+  double skewX;
+  //   horizontal translation
+  double transX;
+  //    vertical skew factor
+  double skewY;
+  //   vertical scale factor
+  double scaleY;
+  //   vertical translation
+  double transY;
+  //    input x-axis perspective factor
+  double pers0;
+  //    input y-axis perspective factor
+  double pers1;
+  //    perspective scale factor
+  double pers2;
+} FlutterTransformation;
+
 typedef bool (*BoolCallback)(void* /* user data */);
+typedef FlutterTransformation (*TransformationCallback)(void* /* user data */);
 typedef uint32_t (*UIntCallback)(void* /* user data */);
+typedef void* (*ProcResolver)(void* /* user data */, const char* /* name */);
 
 typedef struct {
   // The size of this struct. Must be sizeof(FlutterOpenGLRendererConfig).
@@ -41,7 +64,22 @@ typedef struct {
   BoolCallback clear_current;
   BoolCallback present;
   UIntCallback fbo_callback;
+  // This is an optional callback. Flutter will ask the emebdder to create a GL
+  // context current on a background thread. If the embedder is able to do so,
+  // Flutter will assume that this context is in the same sharegroup as the main
+  // rendering context and use this context for asynchronous texture uploads.
+  // Though optional, it is recommended that all embedders set this callback as
+  // it will lead to better performance in texture handling.
   BoolCallback make_resource_current;
+  // By default, the renderer config assumes that the FBO does not change for
+  // the duration of the engine run. If this argument is true, the
+  // engine will ask the embedder for an updated FBO target (via an fbo_callback
+  // invocation) after a present call.
+  bool fbo_reset_after_present;
+  // The transformation to apply to the render target before any rendering
+  // operations. This callback is optional.
+  TransformationCallback surface_transformation;
+  ProcResolver gl_proc_resolver;
 } FlutterOpenGLRendererConfig;
 
 typedef struct {

